### ISSUE TITLE
JDK 11: Async channel closing semantics changed

### ISSUE THREAD

**Issue by jrudolph** [akka/akka#25913]
As observed in #25792, #25882, #25739, and #25733, closing semantics of SocketChannels have changed in JDK 11 (Oracle + OpenJDK).

This is described in the [JDK 11 release notes](https://www.oracle.com/technetwork/java/javase/11-relnote-issues-5012449.html#JDK-8198562):

> Closing a connected SocketChannel that is registered with a Selector will now consistently delay closing the underlying connection until the closed channel is flushed from all selectors that it is registered with. Similarly, closing a ServerSocketChannel that is registered with Selector will now consistently delay closing the underlying listener socket until it is flushed from all selectors that it is registered with. The behavior varied by platform in previous releases. Developers using tools such as netstat to monitor network connection should be aware of this change, particularly with libraries or applications that do not perform selection operations in a timely manner to flush closed channels from selectors.

Before this change the following guarantee was given (at least de facto on Linux):

When the call from `SocketChannel.close` returned:

 1. The JVM part of the socket really was closed
 2. The OS socket was closed (as observable by netstat)
 3. On localhost also the peer had reliably detected closing of the socket (over the network this isn't true necessarily).

Additionally, we had the invariant that the TcpConnection was stopped completely only after the channel had been closed. This was implemented using its `postStop` and could be observed using actor watching. In other words, when you received a `Terminated` for a connection actor you knew that the associated channel was closed before.

We (and potentially also user code) relies on this for several purposes:

 * In Akka IO, we immediately return an acknowledgement message after closing (`Unbound` for listening sockets, one of the `CloseEvent` for connected sockets).
 * Especially for listening sockets the assumption is that once you have received `Unbound` you can rebind the socket (needs guarantee 2).
 * In tests we often rely on guarantee 3 to check the effect of closing a connection.

With JDK 11, the behavior has changed and the (asynchronous) channel is effectively only closed after it has been removed from its `Selector` which might only happen when a call from `selector.select` has returned the next time.

We need to decide which part of the semantics to retain also for JDK 11. Especially hard to maintain is the invariant that closing of a channel happens-before terminating the associated TcpConnection actor because the closing code may happen only in `postStop` where we cannot synchronously block until the selector has finally completed the channel closing.

**Comment by jrudolph** [comment]
IMO we should try to keep as much semantics as possible for JDK 11 users to avoid surprises. In any case, it seems non-trivial to get right for both JDK 8 and JDK 11.

**Comment by jrudolph** [comment]
> Especially hard to maintain is the invariant that closing of a channel happens-before terminating the associated TcpConnection actor because the closing code may happen only in `postStop` where we cannot synchronously block until the selector has finally completed the channel closing.

We could either 

 * make sure to stop the actor only after the SelectionHandler has actually flushed the socket (this doesn't work when the actor just failed and we only notice in `postStop` which would be a bug itself)
 * or, just not provide the invariant and keep the simple code that closes down the channel in postStop, asynchronously pushing the notifications, potentially concurrently with running postStop.

**Comment by patriknw** [comment]
Sounds like the first choice would be good since it has the nice properties for the normal case. If the actor stops because of failure we will only best effort close (asynchronously).

**Comment by jrudolph** [comment]
As @johanandren noted in https://github.com/akka/akka/issues/25733#issuecomment-429780212, a related change is that the semantics of triggering RST using `setSoLinger(0)` has changed. See http://mail.openjdk.java.net/pipermail/nio-dev/2018-May/005101.html and https://bugs.java.com/view_bug.do?bug_id=8203059.

**Comment by jrudolph** [comment]
Turns out the situation is weirder than thought. 

It seems

> > Closing a connected SocketChannel that is registered with a Selector will now consistently delay closing the underlying connection until the closed channel is flushed from all selectors that it is registered with. 

does actually *not* mean that sending out FIN is delayed (for normal connection closure) but that just completely closing the socket is delayed. I.e. the OS `shutdown` call which sends out FIN is not delayed but actually calling `close` later (which will release the OS socket) is delayed.

For RST/setSoLinger(0), however, the situation is exactly as described and sending out of the RST packet is delayed until `select` is called or returning. We have already solved that case by adding a `registration.cancel` call in TcpConnection abort for Windows in #19611. With JDK 11, all platforms now show the same behavior. 

So, what remains for fixing?

 * [ ] Fix TcpConnectionSpec which doesn't use the usual SelectionHandler but registers channels to another Selector
 * [ ] Fix listeners to make sure to wake up SelectionHandler
